### PR TITLE
More improvements for braid groups involving permutations, and a 2-generator presentation

### PR DESCRIPTION
**Pull request by enriqueartal**
This PR is a continuation of #35214.
- Given an element of a symmetric group, the associated permutation braid can be constructed, added in `_element_constructor`.
- The method `_standard_lift_Tietze` has been simplified (for a permutation) and now it gives a shortest word for the lift but not necesarily the smallest for the lexicographic order. The former code has been commented.
- The method `presentation2gens` provides a presentation of the braid group with two generators. The general method `epimorphism` has been adapted to braid groups using this presentation and it is much faster.

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: src/sage/groups/artin.py [hunk 1]
```diff
@@ -103,31 +104,63 @@ def coxeter_group_element(self, W=None):
         Return the corresponding Coxeter group element under the natural
         projection.
 
+        INPUT:
+
+        - W -- (default: ``None``) It may be ``G = SymmetricGroup(n)`` if ``self.parent()``
+          is either an Artin group of type ``['A', n-1]`` or ``BraidGroup(n)`` to obtain an
+          element of ``G`` instead a ``Standard Permutation``.
```
Comment (tscrim) [inline src/sage/groups/artin.py]:
```suggestion
        - ``W`` -- (default: ``self.parent().coxeter_group()``) the image Coxeter group
```

File: src/sage/groups/artin.py [hunk 2]
```diff
@@ -103,31 +104,63 @@ def coxeter_group_element(self, W=None):
...
+
         OUTPUT:
 
-        A permutation.
+        An element of the Coxeter group or an element of a symmetric group.
```
Comment (tscrim) [inline src/sage/groups/artin.py]:
```suggestion
        An element of the Coxeter group ``W``.
```

File: src/sage/groups/artin.py [hunk 3]
```diff
@@ -103,31 +104,63 @@ def coxeter_group_element(self, W=None):
...
+            (1,4,3,2)
+            sage: A.<s1,s2,s3> = BraidGroup(4)
+            sage: c = s1 * s2 * s3^-1
+            sage: c0 = c.coxeter_group_element(); c0
+            [4, 1, 2, 3]
+            sage: c1 = c.coxeter_group_element(W=SymmetricGroup(4)); c1
+            (1,4,3,2)
+
+        From an element of the Coxeter group it is possible to recover
+        the image by the standard section to the Artin group.
```
Comment (tscrim) [inline src/sage/groups/artin.py]:
```suggestion
        the image by the standard section to the Artin group::
```

File: src/sage/groups/artin.py [hunk 4]
```diff
@@ -436,13 +470,13 @@ def __init__(self, coxeter_matrix, names):
         free_group = FreeGroup(names)
         rels = []
         # Generate the relations based on the Coxeter graph
-        I = coxeter_matrix.index_set()
-        for ii,i in enumerate(I):
-            for j in I[ii+1:]:
-                m = coxeter_matrix[i,j]
+        In = coxeter_matrix.index_set()
+        for ii, i in enumerate(In):
+            for j in In[ii+1:]:
```
Comment (tscrim) [inline src/sage/groups/artin.py]: Please ignore (IMO stupid) suggestions regarding nonessential code style that moves it further from common mathematical notation.

File: src/sage/groups/braid.py [hunk 5]
```diff
@@ -439,18 +444,26 @@ def permutation(self, W=None):
 
         OUTPUT:
 
-        The image of ``self`` under the natural projection map to ``W``.
+        The image of ``self`` under the natural projection map to ``W``
+        as either a standard permutation or an element of a symmetric group.
```
Comment (tscrim) [inline src/sage/groups/braid.py]: Revert this change as the symmetric group is an example of the Coxeter group `W` and this extra (unnecessary) verbosity makes it more confusing.

File: src/sage/groups/braid.py [hunk 6]
```diff
@@ -439,18 +444,26 @@ def permutation(self, W=None):
...
+            sage: c0 = b.permutation(W=S); c0
             (1,4,2)
-            sage: c = b.permutation(W=Permutations(4)); c
+            sage: c1 = b.permutation(W=Permutations(4)); c1
             [4, 1, 3, 2]
-            sage: c == b.permutation()
+            sage: c1 == b.permutation()
+            True
+
+        From a permutation it is also possible to recover the permutation braid.
```
Comment (tscrim) [inline src/sage/groups/braid.py]:
```suggestion
        From a permutation it is also possible to recover the permutation braid::
```

File: src/sage/groups/braid.py [hunk 7]
```diff
@@ -2112,12 +2125,11 @@ def deformed_burau_matrix(self, variab='q'):
...
-                                  + [f'{s}m_{i}'
-                                     for i in range(m) if self.Tietze()[i] < 0
-                                     for s in 'bca'])
+        alg = FreeAlgebra(R, m*3,
+                          [f'{s}p_{i}' for i in range(m) if self.Tietze()[i] > 0
+                           for s in 'bca']
+                          + [f'{s}m_{i}' for i in range(m) if self.Tietze()[i] < 0
+                             for s in 'bca'])
         gen_indices = ([i for i in range(m) if self.Tietze()[i] > 0]
                        + [i for i in range(m) if self.Tietze()[i] < 0])
```
Comment (tscrim) [inline src/sage/groups/braid.py]: I would pull out `self.Tietze()` into a separate variable so it is not repeatedly called in these loops.

File: src/sage/groups/braid.py [hunk 8]
```diff
@@ -2281,6 +2293,7 @@ class RightQuantumWord:
          q*cp_1*ap_1 + q^2*bp_1*cm_0*am_0*bm_2
          reduced from ap_1*cp_1 + q^3*bm_2*bp_1*am_0*cm_0
     """
+
```
Comment (tscrim) [inline src/sage/groups/braid.py]: Remove please as the general style is to not have these blanklines.

File: src/sage/groups/braid.py [hunk 9]
```diff
@@ -2805,7 +2828,7 @@ def _links_gould_representation(self, symbolics=False):
             LR = LaurentPolynomialRing(ZZ, 's0r, s1r')
             PR = PolynomialRing(LR, 'Yr')
             s0r, s1r, Yr = PR.gens_dict_recursive().values()
-            pqr = Yr**2 + (s0r**2-1)*(s1r**2 -1)
+            pqr = Yr**2 + (s0r**2-1)*(s1r**2 - 1)
```
Comment (tscrim) [inline src/sage/groups/braid.py]:
```suggestion
            pqr = Yr**2 + (s0r**2-1) * (s1r**2 - 1)
```
PEP8

File: src/sage/groups/braid.py [hunk 10]
```diff
@@ -2901,58 +2924,58 @@ def _LKB_matrix_(self, braid, variab):
             for i in braid[1:]:
                 A = A*self._LKB_matrix_((i,), variab)
             return A
-        l = list(Set(range(n)).subsets(2))
+        n2 = list(Set(range(n)).subsets(2))
```
Comment (tscrim) [inline src/sage/groups/braid.py]: This is probably better to be replaced by `list(itertools.combinations(range(n), 2))`

File: src/sage/groups/braid.py [hunk 11]
```diff
@@ -3397,6 +3420,95 @@ def mirror_involution(self):
...
         return self.hom(gens_mirr, check=False)
 
+    def presentation2gens(self, isomorphisms=False):
+        """
+        Construct an isomorphic group to ``self``, braid group,  with only two generators
+
+        INPUT:
+
+        - ``isomorphism`` -- boolean (default ``False``). If ``True`` an isomorphism
+          from ``self`` and the isomorphic group and its inverse are provided
```
Comment (tscrim) [inline src/sage/groups/braid.py]:
```suggestion
        - ``isomorphism`` -- boolean (default: ``False``); if ``True``, then an isomorphism
          from ``self`` and the isomorphic group and its inverse is also returned
```

File: src/sage/groups/braid.py [hunk 12]
```diff
@@ -3407,7 +3519,7 @@ def BraidGroup(n=None, names='s'):
     - ``n`` -- integer or ``None`` (default). The number of
       strands. If not specified the ``names`` are counted and the
       group is assumed to have one more strand than generators.
-
++
```
Comment (tscrim) [inline src/sage/groups/braid.py]: Please revert this entire change.

File: src/sage/groups/braid.py [hunk 13]
```diff
@@ -3526,6 +3638,7 @@ class group of the punctured disk.
         sage: A(x1^-1, s1)
         x1*x2^-1*x1^-1
     """
+
```
Comment (tscrim) [inline src/sage/groups/braid.py]: Remove; same as before.

File: src/sage/groups/braid.py [hunk 14]
```diff
@@ -3397,6 +3420,95 @@ def mirror_involution(self):
         gens_mirr = [~g for g in self.gens()]
         return self.hom(gens_mirr, check=False)
 
+    def presentation2gens(self, isomorphisms=False):
+        """
+        Construct an isomorphic group to ``self``, braid group,  with only two generators
```
Comment (tscrim) [inline src/sage/groups/braid.py]:
```suggestion
    def presentation_two_generators(self, isomorphisms=False):
        r"""
        Construct a finitely presented group isomorphic to ``self`` with only two generators.
```

File: src/sage/groups/braid.py [hunk 15]
```diff
@@ -3397,6 +3420,95 @@ def mirror_involution(self):
...
+
+            sage: B = BraidGroup(5)
+            sage: B.epimorphisms(SymmetricGroup(5))
+            [Generic morphism:
+            From: Braid group on 5 strands
+            To:   Symmetric group of order 5! as a permutation group
+            Defn: s0 |--> (1,5)
+                    s1 |--> (4,5)
+                    s2 |--> (3,4)
+                    s3 |--> (2,3)]
```
Comment (tscrim) [inline src/sage/groups/braid.py]:
```suggestion
        through the :meth:`two generator presentation
        <presentation_two_generators>` of ``self``.

        INPUT:

        - `H` -- another group

        EXAMPLES::

            sage: B = BraidGroup(5)
            sage: B.epimorphisms(SymmetricGroup(5))
            [Generic morphism:
              From: Braid group on 5 strands
              To:   Symmetric group of order 5! as a permutation group
              Defn: s0 |--> (1,5)
                      s1 |--> (4,5)
                      s2 |--> (3,4)
                      s3 |--> (2,3)]
```

File: src/sage/groups/braid.py (hunk 12 above)
Comment (enriqueartal) [reply src/sage/groups/braid.py:3522]: A blank line appears after `class` in other parts

File: src/sage/groups/braid.py (hunk 8 above)
Comment (enriqueartal) [reply src/sage/groups/braid.py]: A blank line appears after `class` in other parts

File: src/sage/groups/braid.py (hunk 12 above)
Comment (enriqueartal) [reply src/sage/groups/braid.py]: I do not understand *this entire change*. There was an extra `+` due to my fingers, removed.

File: src/sage/groups/braid.py (hunk 8 above)
Comment (tscrim) [reply src/sage/groups/braid.py]: They shouldn't though as we do not have them for most classes in Sage (although we aren't yet uniform in this).

File: src/sage/groups/braid.py (hunk 12 above)
Comment (tscrim) [reply src/sage/groups/braid.py]: Blanklines between inputs are unnecessary and evil IMO. I don't want to add them anywhere, especially where it wasn't before. (It also helps prevent merge conflicts as it is not where you are making changes. Many of your other changes fall under this too, but I don't disapprove of them.)

File: src/sage/groups/braid.py (hunk 10 above)
Comment (enriqueartal) [reply src/sage/groups/braid.py]: I made only style changes. I agree for the change but I prefer to keep as it is since for the sequel it is important for the elements of `n2` to be sets.

Comment (tscrim) [reply src/sage/groups/braid.py]: Do you mean a Sage `Set` or is a Python `set` sufficient? It is easy enough to cast them as
```python
[set(X) for X in itertools.combinations(range(n), 2)]
```
and this should be faster (although it is less obvious to a non-Python programmer what the line is doing).

File: src/sage/groups/braid.py (hunk 12 above)
Comment (enriqueartal) [reply src/sage/groups/braid.py]: I reverted both changes, but actually in this file for most classes there is a blank line (not introduced by me)

File: src/sage/groups/braid.py (hunk 10 above)
Comment (enriqueartal) [reply src/sage/groups/braid.py]: Changed, but I get an error if I do not import it previously, so I added at the beginning `from itertools import combinations` and remove `itertools` at this place.

**Comment by enriqueartal** [comment]
Now the method `epimorphisms` returns a list of generic morphisms. I guess it was the only choice when this method was created, but now group homomorphisms exist. I wonder if it would be better to change it. In that case some `GAP` functions would be accessible.

File: src/sage/groups/braid.py (hunk 10 above)
Comment (tscrim) [reply src/sage/groups/braid.py]: Sorry, I should have been explicit about the import (I am used to always having to do it).

**Comment by tscrim** [comment]
It doesn't seem like the homspace knows how to make itself using the more specific class. I would put that off for another ticket.

File: src/sage/groups/artin.py [hunk 16]
```diff
@@ -47,6 +47,7 @@ class ArtinGroupElement(FinitelyPresentedGroupElement):
         sage: A((1, 2, -3, -2))
         s1*s2*s3^-1*s2^-1
     """
+
```
Comment (tscrim) [inline src/sage/groups/artin.py]: Really,please remove this blankline..

File: src/sage/groups/artin.py [hunk 17]
```diff
@@ -103,31 +104,61 @@ def coxeter_group_element(self, W=None):
...
-        return W.prod(s[I[abs(i)-1]] for i in self.Tietze())
+        In = W.index_set()
+        return W.prod(s[In[abs(i)-1]] for i in self.Tietze())
+
 
 class FiniteTypeArtinGroupElement(ArtinGroupElement):
     """
     An element of a finite-type Artin group.
     """
+
```
Comment (tscrim) [inline src/sage/groups/artin.py]: Same here.

File: src/sage/groups/artin.py [hunk 18]
```diff
@@ -695,6 +727,7 @@ class FiniteTypeArtinGroup(ArtinGroup):
         sage: GF = F.cayley_graph(elements=ball(F, 4), generators=F.gens()); GF
         Digraph on 40 vertices
     """
+
```
Comment (tscrim) [inline src/sage/groups/artin.py]: And here.

File: src/sage/groups/braid.py [hunk 19]
```diff
@@ -112,6 +117,7 @@ class Braid(FiniteTypeArtinGroupElement):
         sage: B((1, 2, -3, -2))
         s0*s1*s2^-1*s1^-1
     """
+
```
Comment (tscrim) [inline src/sage/groups/braid.py]: And here.

File: src/sage/groups/braid.py [hunk 20]
```diff
@@ -2732,30 +2750,33 @@ def _standard_lift_Tietze(self, p):
...
+        # while pl.length():
+        #     i = 1
+        #     while i < max(pl):
+        #         if pl(i) > pl(i+1):
+        #             l.append(i)
+        #             pl = self._coxeter_group.simple_reflection(i) * pl
+        #             i = 1
+        #         else:
+        #             i += 1
+        # return tuple(l)
```
Comment (tscrim) [inline src/sage/groups/braid.py]: You can remove the old code. It is there in the git history.

File: src/sage/groups/braid.py [hunk 21]
```diff
@@ -3397,6 +3418,96 @@ def mirror_involution(self):
...
+    def presentation_two_generators(self, isomorphisms=False):
+        r"""
+        Construct a finitely presented group isomorphic to ``self`` with only two generators.
+
+        INPUT:
+
+        - ``isomorphism`` -- boolean (default ``False``); if ``True``, then an isomorphism
+          from ``self`` and the isomorphic group and its inverse is also returned
+
+        EXAMPLES:
```
Comment (tscrim) [inline src/sage/groups/braid.py]:
```suggestion
        EXAMPLES::
```

File: src/sage/groups/braid.py [hunk 22]
```diff
@@ -446,11 +452,18 @@ def permutation(self, W=None):
...
+            sage: c0 = b.permutation(W=S); c0
             (1,4,2)
-            sage: c = b.permutation(W=Permutations(4)); c
+            sage: c1 = b.permutation(W=Permutations(4)); c1
             [4, 1, 3, 2]
-            sage: c == b.permutation()
+            sage: c1 == b.permutation()
+            True
+
+        From a permutation it is also possible to recover the permutation braid::
```
Comment (tscrim) [inline src/sage/groups/braid.py]:
```suggestion
        It also applies the canonical section to a permutation::
```
The current version implies a certain uniqueness that is simply not true. Consider $s^3$ versus $s$ for a simple transposition $s$.

**Review by tscrim — CHANGES_REQUESTED** [review]
Just a few more details. A number of your changes are not really useful (e.g., reordering the import statements in some arbitrary order (yes, I know it is alphabetical, but that is still an arbitrary choice)) that could create trivial conflicts with other changes, but it's already done. However, the stylistic changes will not get a positive review from me.

File: src/sage/groups/braid.py (hunk 22 above)
Comment (enriqueartal) [reply src/sage/groups/braid.py]: I put another sentence. I really want to keep the expression `permutation braid` (or `positive permutation braid`) since is a common concept in braid group theory. Please check it.

**Comment by enriqueartal** [comment]
I applied you comments. One should be reviewed again. Probably not now, by I think that in general `epimorphisms` should give group homomorphisms and not generic ones.

File: src/sage/groups/braid.py [hunk 23]
```diff
@@ -446,11 +451,20 @@ def permutation(self, W=None):
...
-            sage: c = b.permutation(W=Permutations(4)); c
+            sage: c1 = b.permutation(W=Permutations(4)); c1
             [4, 1, 3, 2]
-            sage: c == b.permutation()
+            sage: c1 == b.permutation()
+            True
+
+        The canonical section from the symmetric group to the braid group
+        (sending a permutation to its associated permutation braid)
+        can be recovered.::
```
Comment (tscrim) [inline src/sage/groups/braid.py]:
```suggestion
        The canonical section from the symmetric group to the braid group
        (sending a permutation to its associated permutation braid)
        can be recovered::
```
I don't know that area of the literature as well as you, but I hadn't seen that terminology before. This is also better to the novice with "associated" added. Thank you. Anyways, just need to remove the period/full-stop as the outputted doc is "can be recovered.:"

**Review by tscrim — APPROVED** [review]
Thank you. I am going to approve this PR; just make the one last change and then I will set this to a positive review.

**Comment by tscrim** [comment]
Thank you.
